### PR TITLE
 bpo-32571: Avoid raising unneeded AttributeError and silencing it in C code (alt).

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
Add `_PyObject_LookupAttr()` and `_PyObject_LookupAttrId()` and use them in appropriate cases instead of `PyObject_GetAttr()`, `_PyObject_GetAttrId()` and `_PyObject_GetAttrIdWithoutError()`.

https://bugs.python.org/issue32571

### CODE REVIEW

File: Include/object.h [hunk 1]
```diff
@@ -536,11 +536,20 @@ PyAPI_FUNC(int) PyObject_SetAttr(PyObject *, PyObject *, PyObject *);
...
+/* Replacements of PyObject_GetAttr() and _PyObject_GetAttrId() which
+   don't raise AttributeError.
+
+   Return 1 and set *result != NULL if an attribute is found.
+   Return 0 and set *result == NULL if an attribute is not found;
+   an AttributeError is silenced.
+   Return -1 and set *result == NULL if an error other than AttributeError
+   is raised.
+*/
+PyAPI_FUNC(int) _PyObject_LookupAttr(PyObject *, PyObject *, PyObject **);
```
Comment (1st1) [inline Include/object.h:551]: I would rename the method to `PyObject_LookupAttr` (no leading underscore).  I'd use it in my libraries and the leading underscore would make me think that I shouldn't.

I also like Inada-san's suggestion to simplify the return value even further:

* `-1`, if an *unrelated* error occurred during attribute lookup
* `0` if no errors have occurred; `*result` will be set to `NULL` if the attribute was not found, and to a non-`NULL` value if it was found.

Comment (serhiy-storchaka) [reply Include/object.h:551]: I think this API is experimental and should be kept private for some time. We are still discussing it, and after making some design decision we may change it in future.

Comment (methane) [reply Include/object.h:551]: I agree with @serhiy-storchaka.
While I feel this API is nice, I'm not sure we should set
`*result=NULL` when returning -1 or not.

Comment (1st1) [reply Include/object.h:551]: Don't we care C API methods with a leading underscore the same as without?  I've never seen us breaking the `_` prefixed API to be honest. So leading underscore or not it's set in stone.

**Review by methane — APPROVED** [review]

File: Include/object.h (hunk 1 above)
Comment (methane) [reply Include/object.h:551]: @1st1 Actually speaking, I broke `_PyObject_GenericGetAttrWithDict()`.  I added one argument to suppress AttributeError.

Comment (1st1) [reply Include/object.h:551]: Fair enough. The PR is LGTM.

**Review by 1st1 — APPROVED** [review]

**Comment by methane** [comment]
@serhiy-storchaka Would you merge this?
I want to use this API in #5273.

**Pull request merged**